File: modules/widgets/lib/rin/widgets/ContainerWidget.cpp
#include "rin/widgets/ContainerWidget.hpp"

#include "rin/widgets/ContainerWidgetSlot.hpp"
#include "rin/widgets/event/CursorDownEvent.hpp"
#include "rin/widgets/event/CursorMoveEvent.hpp"
#include "rin/widgets/event/ScrollEvent.hpp"
#include "rin/widgets/graphics/WidgetDrawCommands.hpp"

void ContainerWidget::OnChildResized(Widget* widget)
{
    if (CheckSize()) ArrangeSlots(GetContentSize());
}

void ContainerWidget::OnChildSlotUpdated(ContainerWidgetSlot* slot)
{
    if (CheckSize()) ArrangeSlots(GetContentSize());
}

Shared<ContainerWidgetSlot> ContainerWidget::MakeSlot(const Shared<Widget>& widget)
{
    return newShared<ContainerWidgetSlot>(this, widget);
}

ContainerWidget::ContainerWidget()
{
    SetVisibility(WidgetVisibility::VisibleNoHitTestSelf);
}

size_t ContainerWidget::GetMaxSlots() const
{
    return std::numeric_limits<size_t>::max();
}

size_t ContainerWidget::GetUsedSlots() const
{
    return _slots.size();
}

Shared<ContainerWidgetSlot> ContainerWidget::GetSlot(int index) const
{
    return index < _slots.size() ? _slots[index] : Shared<ContainerWidgetSlot>{};
}

Shared<ContainerWidgetSlot> ContainerWidget::GetSlot(Widget* widget) const
{
    return _widgetsToSlots.contains(widget) ? _widgetsToSlots.at(widget) : Shared<ContainerWidgetSlot>{};
}

std::vector<Shared<ContainerWidgetSlot>> ContainerWidget::GetSlots() const
{
    return _slots;
}

Shared<ContainerWidgetSlot> ContainerWidget::AddChild(const Shared<Widget>& widget)
{
    if (!widget) return {};

    auto newSlotCount = _slots.size() + 1;

    if (newSlotCount > GetMaxSlots()) return {};

    auto slot = MakeSlot(widget);
    _slots.push_back(slot);
    _widgetsToSlots.emplace(widget.get(), slot);
    widget->SetParent(this->GetSharedDynamic<ContainerWidget>());

    if (auto surf = GetSurface())
    {
        widget->NotifyAddedToSurface(surf);
        OnChildResized(widget.get());
    }

    return slot;
}

bool ContainerWidget::RemoveChild(const Shared<Widget>& widget)
{
    if (!_widgetsToSlots.contains(widget.get())) return false;

    for (auto i = 0; i < _slots.size(); i++)
    {
        if (_slots[i]->GetWidget() != widget) continue;

        widget->SetParent({});

        auto surf = GetSurface();

        if (surf)
        {
            widget->NotifyRemovedFromSurface(surf);
        }

        _slots.erase(_slots.begin() + i);
        _widgetsToSlots.erase(widget.get());

        if (surf)
        {
            CheckSize();
        }

        return true;
    }

    return false;
}

void ContainerWidget::SetSize(const Vec2<float>& size)
{
    Widget::SetSize(size);
    if (auto surf = GetSurface())
    {
        ArrangeSlots(GetContentSize());
    }
}

void ContainerWidget::OnDispose(bool manual)
{
    Widget::OnDispose(manual);

    for (auto& slot : GetSlots())
    {
        slot->GetWidget()->Dispose();
    }

    _slots.clear();
}

void ContainerWidget::NotifyAddedToSurface(const Shared<WidgetSurface>& widgetSurface)
{
    Widget::NotifyAddedToSurface(widgetSurface);
    for (auto& slot : GetSlots())
    {
        slot->GetWidget()->NotifyAddedToSurface(widgetSurface);
    }
}

void ContainerWidget::NotifyRemovedFromSurface(const Shared<WidgetSurface>& widgetSurface)
{
    Widget::NotifyRemovedFromSurface(widgetSurface);
    for (auto& slot : GetSlots())
    {
        slot->GetWidget()->NotifyRemovedFromSurface(widgetSurface);
    }
}

Shared<Widget> ContainerWidget::NotifyCursorDown(const Shared<CursorDownEvent>& event, const TransformInfo& transform)
{
    if (AreChildrenHitTestable())
    {
        if (auto widget = NotifyChildrenCursorDown(event, transform)) return widget;
    }
    return Widget::NotifyCursorDown(event, transform);
}

void ContainerWidget::NotifyCursorEnter(const Shared<CursorMoveEvent>& event, const TransformInfo& transform,
                                        std::vector<Shared<Widget>>& items)
{
    if (AreChildrenHitTestable()) NotifyChildrenCursorEnter(event, transform, items);

    Widget::NotifyCursorEnter(event, transform, items);
}

bool ContainerWidget::NotifyCursorMove(const Shared<CursorMoveEvent>& event, const TransformInfo& transform)
{
    if (AreChildrenHitTestable() && NotifyChildrenCursorMove(event, transform)) return true;
    return Widget::NotifyCursorMove(event, transform);
}

bool ContainerWidget::NotifyScroll(const Shared<ScrollEvent>& event, const TransformInfo& transform)
{
    if (AreChildrenHitTestable() && NotifyChildrenScroll(event, transform)) return true;
    return Widget::NotifyScroll(event, transform);
}

Shared<Widget> ContainerWidget::NotifyChildrenCursorDown(const Shared<CursorDownEvent>& event,
                                                         const TransformInfo& transform)
{
    for (auto& slot : GetSlots())
    {
        if (auto slotTransform = ComputeChildTransform(slot, transform); slotTransform.IsPointWithin(event->position) && slot->GetWidget()->NotifyCursorDown(event, slotTransform))
            return slot->GetWidget();
    }
    return {};
}

bool ContainerWidget::NotifyChildrenCursorEnter(const Shared<CursorMoveEvent>& event, const TransformInfo& transform,
                                                std::vector<Shared<Widget>>& items)
{
    for (auto& slot : GetSlots())
    {
        if (auto slotTransform = ComputeChildTransform(slot, transform); slotTransform.IsPointWithin(event->position))
        {
            slot->GetWidget()->NotifyCursorEnter(event, slotTransform, items);
        }
    }
    return false;
}

bool ContainerWidget::NotifyChildrenCursorMove(const Shared<CursorMoveEvent>& event, const TransformInfo& transform)
{
    for (auto& slot : GetSlots())
    {
        if (auto slotTransform = ComputeChildTransform(slot, transform); slotTransform.IsPointWithin(event->position) && slot->GetWidget()->NotifyCursorMove(event, slotTransform))
            return true;
    }
    return false;
}

bool ContainerWidget::NotifyChildrenScroll(const Shared<ScrollEvent>& event, const TransformInfo& transform)
{
    for (auto& slot : GetSlots())
    {
        if (auto slotTransform = ComputeChildTransform(slot, transform); slotTransform.IsPointWithin(event->position) && slot->GetWidget()->NotifyScroll(event, slotTransform))
            return true;
    }
    return false;
}

void ContainerWidget::CollectContent(const TransformInfo& transform, WidgetDrawCommands& drawCommands)
{
    const auto slots = GetSlots();
    const auto clipMode = GetClipMode();
    if (clipMode == EClipMode::None)
    {
        for (auto& slot : slots)
        {
            TransformInfo t = ComputeChildTransform(slot, transform);
            const auto widget = slot->GetWidget();
            widget->Collect(t, drawCommands);
        }
    }
    else if (clipMode == EClipMode::Bounds)
    {
        drawCommands.PushClip(transform, this);

        const auto myAABR = transform.ComputeAxisAlignedBoundingRect();

        for (auto& slot : slots)
        {
            TransformInfo t = ComputeChildTransform(slot, transform);
            const auto slotAABR = t.ComputeAxisAlignedBoundingRect();

            if (!myAABR.IntersectsWith(slotAABR)) continue;

            const auto widget = slot->GetWidget();
            widget->Collect(t, drawCommands);
        }

        drawCommands.PopClip();
    }
}

TransformInfo ContainerWidget::ComputeChildTransform(const Shared<ContainerWidgetSlot>& slot,
                                                     const TransformInfo& myTransform)
{
    return ComputeChildTransform(slot->GetWidget(), myTransform);
}

TransformInfo ContainerWidget::ComputeChildTransform(const Shared<Widget>& widget, const TransformInfo& myTransform)
{
    auto padding = GetPadding();
    return TransformInfo{
        myTransform.transform.Translate(Vec2(padding.left, padding.top)) * widget->ComputeRelativeTransform(),
        widget->GetSize(), myTransform.depth + 1
    };
}

EClipMode ContainerWidget::GetClipMode() const
{
    return _clipMode;
}

void ContainerWidget::SetClipMode(EClipMode clipMode)
{
    _clipMode = clipMode;
}

File: rin.Widgets/ContainerWidget.cs
using System.Collections.Concurrent;
using rin.Core.Animation;
using rin.Core.Extensions;
using rin.Core.Math;
using rin.Widgets.Enums;
using rin.Widgets.Events;
using rin.Widgets.Graphics;
using rin.Widgets.Graphics.Commands;

namespace rin.Widgets;

public abstract class ContainerWidget : Widget
{
    private readonly ConcurrentDictionary<Widget, ContainerSlot> _widgetSlotMap = [];
    private readonly List<ContainerSlot> _slots = [];

    protected readonly Mutex SlotsMutex = new();

    public Clip Clip = Clip.None;


    /// <summary>
    /// Adds the Widget to this container
    /// </summary>
    public Widget Child
    {
        init => AddChild(value);
    }

    /// <summary>
    /// Adds the widgets to this container
    /// </summary>
    public Widget[] Children
    {
        init
        {
            foreach (var slot in value)
            {
                AddChild(slot);
            }
        }
    }

    /// <summary>
    /// Adds the slot to this container
    /// </summary>
    public ContainerSlot Slot
    {
        init => AddChild(value);
    }


    /// <summary>
    /// Adds the slots to this container
    /// </summary>
    public ContainerSlot[] Slots
    {
        init
        {
            foreach (var slot in value)
            {
                AddChild(slot);
            }
        }
    }

    /// <summary>
    /// Arranges content and returns their computed total size i.e. the combined length of all items in a list
    /// </summary>
    /// <param name="availableSpace"></param>
    /// <returns></returns>
    protected abstract Vector2<float> ArrangeContent(Vector2<float> availableSpace);


    protected virtual ContainerSlot MakeSlot(Widget widget)
    {
        return new ContainerSlot(this)
        {
            Child = widget
        };
    }

    public virtual ContainerSlot? AddChild<TE>() where TE : Widget
    {
        return AddChild(Activator.CreateInstance<TE>());
    }

    public virtual void OnChildInvalidated(Widget child, InvalidationType invalidation)
    {
        if (_widgetSlotMap.TryGetValue(child, out var slot))
        {
            OnSlotInvalidated(slot, invalidation);
        }
    }

    public virtual void OnSlotInvalidated(ContainerSlot slot, InvalidationType invalidation)
    {
        Invalidate(invalidation);
    }

    protected override Vector2<float> LayoutContent(Vector2<float> availableSpace)
    {
        return ArrangeContent(availableSpace);
    }

    public ContainerSlot? AddChild(Widget widget) => AddChild(MakeSlot(widget));

    public ContainerSlot? AddChild(ContainerSlot slot)
    {
        lock (SlotsMutex)
        {
            var maxSlots = GetMaxSlotsCount();
            if (maxSlots > 0 && _slots.Count == maxSlots) return null;

            var widget = slot.Child;

            widget.SetParent(this);

            slot.SetOwner(this);

            _slots.Add(slot);
            _widgetSlotMap.TryAdd(widget, slot);

            if (Surface != null)
            {
                widget.NotifyAddedToSurface(Surface);
                OnSlotInvalidated(slot, InvalidationType.DesiredSize);
            }

            //Console.WriteLine("Added child [{0}] to container [{1}]", widget.GetType().Name, GetType().Name);
            return slot;
        }
    }

    public virtual bool RemoveChild(Widget widget)
    {
        lock (SlotsMutex)
        {
            for (var i = 0; i < _slots.Count; i++)
            {
                if (_slots[i].Child != widget) continue;

                widget.SetParent(null);

                if (Surface != null)
                {
                    widget.NotifyRemovedFromSurface(Surface);
                }

                _slots.RemoveAt(i);
                _widgetSlotMap.TryRemove(widget, out var _);

                if (Surface != null)
                {
                    Invalidate(InvalidationType.DesiredSize);
                }

                return true;
            }

            return false;
        }
    }

    public virtual ContainerSlot? GetSlot(int idx)
    {
        lock (SlotsMutex)
        {
            return idx < _slots.Count ? _slots[idx] : null;
        }
    }

    public virtual ContainerSlot[] GetSlots()
    {
        lock (SlotsMutex)
        {
            return _slots.ToArray();
        }
    }

    public virtual int GetSlotsCount()
    {
        lock (SlotsMutex)
        {
            return _slots.Count;
        }
    }

    public virtual int GetMaxSlotsCount() => int.MaxValue;

    public override Widget? NotifyCursorDown(CursorDownEvent e, Matrix3 transform)
    {
        if (IsChildrenHitTestable)
        {
            var res = ChildrenNotifyCursorDown(e, transform.Translate(new Vector2<float>(Padding.Left, Padding.Top)));
            if (res != null) return res;
        }

        return base.NotifyCursorDown(e, transform);
    }

    protected virtual Widget? ChildrenNotifyCursorDown(CursorDownEvent e, Matrix3 transform)
    {
        foreach (var slot in GetHitTestableSlots().AsReversed())
        {
            var slotTransform = ComputeSlotTransform(slot, transform);
            if (!slot.Child.PointWithin(slotTransform, e.Position)) continue;
            var res = slot.Child.NotifyCursorDown(e, slotTransform);
            if (res != null) return res;
        }

        return null;
    }

    public override void NotifyCursorEnter(CursorMoveEvent e, Matrix3 transform, List<Widget> items)
    {
        if (IsChildrenHitTestable)
            ChildrenNotifyCursorEnter(e, transform.Translate(new Vector2<float>(Padding.Left, Padding.Top)), items);

        base.NotifyCursorEnter(e, transform, items);
    }

    protected virtual void ChildrenNotifyCursorEnter(CursorMoveEvent e, Matrix3 transform, List<Widget> items)
    {
        foreach (var slot in GetHitTestableSlots())
        {
            var slotTransform = ComputeSlotTransform(slot, transform);
            if (slot.Child.PointWithin(slotTransform, e.Position))
                slot.Child.NotifyCursorEnter(e, slotTransform, items);
        }
    }

    public override bool NotifyCursorMove(CursorMoveEvent e, Matrix3 transform)
    {
        if (IsChildrenHitTestable &&
            ChildrenNotifyCursorMove(e, transform.Translate(new Vector2<float>(Padding.Left, Padding.Top))))
            return true;

        return base.NotifyCursorMove(e, transform);
    }

    protected virtual bool ChildrenNotifyCursorMove(CursorMoveEvent e, Matrix3 transform)
    {
        foreach (var slot in GetHitTestableSlots())
        {
            var slotTransform = ComputeSlotTransform(slot, transform);
            if (slot.Child.PointWithin(slotTransform, e.Position) &&
                slot.Child.NotifyCursorMove(e, slotTransform))
                return true;
        }

        return false;
    }

    public override bool NotifyScroll(ScrollEvent e, Matrix3 transform)
    {
        if (IsChildrenHitTestable &&
            ChildrenNotifyScroll(e, transform.Translate(new Vector2<float>(Padding.Left, Padding.Top)))) return true;

        return base.NotifyScroll(e, transform);
    }

    protected virtual bool ChildrenNotifyScroll(ScrollEvent e, Matrix3 transform)
    {
        foreach (var slot in GetHitTestableSlots())
        {
            var slotTransform = ComputeSlotTransform(slot, transform);
            if (slot.Child.PointWithin(slotTransform, e.Position) &&
                slot.Child.NotifyScroll(e, slotTransform))
                return true;
        }

        return false;
    }

    public override void NotifyAddedToSurface(Surface surface)
    {
        base.NotifyAddedToSurface(surface);
        lock (SlotsMutex)
        {
            foreach (var slot in _slots)
                slot.Child.NotifyAddedToSurface(surface);
        }
    }

    public override void NotifyRemovedFromSurface(Surface surface)
    {
        base.NotifyRemovedFromSurface(surface);
        lock (SlotsMutex)
        {
            foreach (var slot in _slots)
                slot.Child.NotifyRemovedFromSurface(surface);
        }
    }

    protected override void OnDispose(bool isManual)
    {
        base.OnDispose(isManual);
        lock (SlotsMutex)
        {
            foreach (var slot in _slots) slot.Child.Dispose();
        }
    }

    /// <summary>
    /// Compute extra offsets for this slot
    /// </summary>
    /// <param name="slot"></param>
    /// <param name="contentTransform"></param>
    /// <returns></returns>
    protected virtual Matrix3 ComputeSlotTransform(ContainerSlot slot, Matrix3 contentTransform)
    {
        var relativeTransform = slot.Child.ComputeRelativeTransform();
        return contentTransform * relativeTransform;
    }

    public virtual IEnumerable<ContainerSlot> GetCollectableSlots() => GetSlots();
    public virtual IEnumerable<ContainerSlot> GetHitTestableSlots() => GetSlots();

    protected virtual void CollectSlots(Matrix3 transform, Rect clip, DrawCommands drawCommands)
    {
    }

    public override void Collect(Matrix3 transform, Rect clip, DrawCommands drawCommands)
    {
        ((IAnimatable)this).Update();
        
        if (Visibility is Visibility.Hidden or Visibility.Collapsed)
        {
            return;
        }

        drawCommands.IncrDepth();
        var clipRect = clip;

        if (Parent != null && Clip == Clip.Bounds)
        {
            drawCommands.PushClip(transform, GetContentSize());
        }

        if (Clip == Clip.Bounds)
        {
            clipRect = ComputeAABB(transform).Clamp(clipRect);
        }

        var transformWithPadding = transform.Translate(new Vector2<float>(Padding.Left, Padding.Top));

        foreach (var slot in GetCollectableSlots())
        {
            var slotTransform = ComputeSlotTransform(slot, transformWithPadding);

            var aabb = slot.Child.ComputeAABB(slotTransform);

            if (!clipRect.IntersectsWith(aabb)) continue;

            slot.Child.Collect(slotTransform, clipRect, drawCommands);
        }

        if (Parent != null && Clip == Clip.Bounds)
        {
            drawCommands.PopClip();
        }
    }
}